pkg/destroy/powervs/serviceinstance.go

```diff
@@ -0,0 +1,314 @@
+}
+
+// listServiceInstances list service instances for the cluster.
+func (o *ClusterUninstaller) listServiceInstances() (cloudResources, error) {
+	o.Logger.Debugf("Listing service instances")
+
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	select {
+	case <-ctx.Done():
+		o.Logger.Debugf("listServiceInstances: case <-ctx.Done()")
+		return nil, o.Context.Err() // we're cancelled, abort
+	default:
+	}
+
+	var (
+		resourceGroupID string
+		options         *resourcecontrollerv2.ListResourceInstancesOptions
+		resources       *resourcecontrollerv2.ResourceInstancesList
+		err             error
+		perPage         int64 = 10
+		moreData              = true
+		nextURL         *string
+	)
+
+	resourceGroupID, err = o.convertResourceGroupNameToID(o.resourceGroupID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to convert resourceGroupID")
+	}
+	o.Logger.Debugf("listServiceInstances: converted %v to %v", o.resourceGroupID, resourceGroupID)
+
+	options = o.controllerSvc.NewListResourceInstancesOptions()
+	// options.SetType("resource_instance")
+	options.SetResourceGroupID(resourceGroupID)
+	options.SetResourceID(virtualServerResourceID)
+	options.SetLimit(perPage)
+
+	result := []cloudResource{}
+
+	for moreData {
+		if options.Start != nil {
+			o.Logger.Debugf("listServiceInstances: options = %+v, options.Limit = %v, options.Start = %v, options.ResourceGroupID = %v", options, *options.Limit, *options.Start, *options.ResourceGroupID)
+		} else {
+			o.Logger.Debugf("listServiceInstances: options = %+v, options.Limit = %v, options.ResourceGroupID = %v", options, *options.Limit, *options.ResourceGroupID)
+		}
+
+		resources, _, err = o.controllerSvc.ListResourceInstancesWithContext(ctx, options)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list resource instances")
+		}
+
+		o.Logger.Debugf("listServiceInstances: resources.RowsCount = %v", *resources.RowsCount)
+
+		for _, resource := range resources.Resources {
+			var (
+				getResourceOptions *resourcecontrollerv2.GetResourceInstanceOptions
+				resourceInstance   *resourcecontrollerv2.ResourceInstance
+				response           *core.DetailedResponse
+			)
+
+			o.Logger.Debugf("listServiceInstances: resource.Name = %s", *resource.Name)
+
+			getResourceOptions = o.controllerSvc.NewGetResourceInstanceOptions(*resource.ID)
+
+			resourceInstance, response, err = o.controllerSvc.GetResourceInstance(getResourceOptions)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to get instance: %s", response)
+			}
+			if response != nil && response.StatusCode == gohttp.StatusNotFound {
+				o.Logger.Debugf("listServiceInstances: gohttp.StatusNotFound")
+				continue
+			} else if response != nil && response.StatusCode == gohttp.StatusInternalServerError {
+				o.Logger.Debugf("listServiceInstances: gohttp.StatusInternalServerError")
+				continue
+			}
+
+			if resourceInstance.Type == nil {
```

Comment:
Is there a scenario where the resourceInstance.Type can be nil?

Reply:
As it's a pointer, yes.